feat(i18n): restructure admin-form directory, add English translations

jednhk opened the pull request:
## Problem
Related to recent PR #7456 opened by @helloitsxinyi.
Issue was discussed with @LoneRifle.

### Context
Currently we have translation for some features, namely “common”, “login”, and “public form” feature translations, found in the i18n directory.

However, the admin form is still missing en-sg translations.

## Solution
This PR adds a new admin-form directory, frontend/src/features/admin-form, containing translations for some of the features within the admin form builder, such as the sidebar elements depicted below.

<img width="794" alt="image" src="https://github.com/opengovsg/FormSG/assets/69101594/efaaa04c-61ab-4869-83ef-80e82b0334ce">

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->
- [ ] Yes - this PR contains breaking changes
    - Details ...
- [x] No - this PR is backwards compatible  

**Features**:

- NA

**Improvements**:
- [x] A new directory, and subdirectories has been created for the locale files and can be found at frontend/src/i18n/locales/features/admin-form
- [x] **(TO DO)** Hardcoded values in relevant files have been replaced with the corresponding translations.

**Bug Fixes**:

- NA

## Screenshots

**BEFORE**:

<img width="127" alt="image" src="https://github.com/user-attachments/assets/a5c08cc3-1626-4c3a-a421-e9631e517acb">

**AFTER**:

<img width="1007" alt="image" src="https://github.com/user-attachments/assets/4ff3be6d-00c5-4f4d-803c-b18a55c9687b">


## Tests
All tests as these changes involve the external facing UI.

jednhk committed:
feat(i18n): add admin-form feature directory

Per discussions, create a nested dir structure for translations
pertaining to the admin form.

- create admin-form under features
  - nest sidebar and navbar, moving the latter from admin-form-navbar
- update types and index to reflect changes

jednhk commented:
# Queries
@KenLSM Below are some queries regarding the newly proposed file structure. Once resolved, I will proceed to replace the hardcoded English strings with the i18n Translation hooks.

## Naming
If you refer to `frontend/src/i18n/locales/types.ts`:
<img width="458" alt="image" src="https://github.com/user-attachments/assets/9ae36fe7-f4aa-4caf-bbd3-92ea7e1a714a">
I'm not sure if I should change all the names to:

- Fields -> AdminFormFields
- HeaderAndInstructions -> AdminFormHeaderAndInstructions (though this is quite long)
- Logic -> AdminFormLogic
- ThankYou -> AdminFormThankYou

I believe it's clearer, but its just quite lengthy.

## Purpose of `admin-form/index.ts` file
In all of the translation directories (e.g. common, login, public-form), the index.ts file is purposed as a **point of re-export** and to **define the translation interface** as such:

<img width="465" alt="image" src="https://github.com/user-attachments/assets/4cef129f-7b0c-4fc6-9f23-5742a8ba0a3e">

However, for my new admin-form directory, it doesn't have any translation types. It only serves to further organize the translations into `admin-form/navbar` and `admin-form/sidebar`.

**As such, do I need an `admin-form/index.ts` file? If so, what should it contain?**

LoneRifle commented on frontend/src/i18n/locales/features/admin-form/sidebar/fields/en-sg.ts:
```diff
@@ -0,0 +1,194 @@
+      placeholder: 'Enter one option per line',
+    },
+    duplicateOptionsError: 'Please remove duplicate options',
+    otherInvalidInputError: 'Harap masukan input lainnya',
```
This isn't English at all. Please find the corresponding English in the codebase by comparison and inspection

jednhk replied:
Thanks for the catch. Made the amendments.

LoneRifle replied:
Have the amendments been made? The text still reflects a language other than English

jednhk replied:
Just pushed the amendments.

LoneRifle replied:
This isn't an amendment at all; you removed the translation altogether. Could you please find the original English text in FormSG? You can do this by identifying the component where this key is used in the fork you were briefed on, and then locating the original text in FormSG.

jednhk replied:
Apologies, I misunderstood the task before. 

Just added the respective translation, it was found in the `RadioField.tsx` file. Do let me know if its good now.

LoneRifle reviewed (changes requested):
Just blocking the PR given the stray mistranslation I saw

LoneRifle commented:
@KenLSM - could you have a look please?

jednhk commented:
Hi @KenLSM, I noticed you've already marked the PR as ready for review. However, I've yet to populate the codebase with the Translation hooks.

Could I clarify if the current directory structure is fine? In which case, I will start to populate the codebase with the new Translation hooks.

Refer to the queries regarding the PR found below.

> # Queries
> @KenLSM Below are some queries regarding the newly proposed file structure. Once resolved, I will proceed to replace the hardcoded English strings with the i18n Translation hooks.
> 
> ## Naming
> If you refer to `frontend/src/i18n/locales/types.ts`: <img alt="image" width="458" src="https://private-user-images.githubusercontent.com/69101594/348059733-9ae36fe7-f4aa-4caf-bbd3-92ea7e1a714a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExMzAyOTksIm5iZiI6MTcyMTEyOTk5OSwicGF0aCI6Ii82OTEwMTU5NC8zNDgwNTk3MzMtOWFlMzZmZTctZjRhYS00Y2FmLWJiZDMtOTJlYTdlMWE3MTRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE2VDExMzk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViYzc0OTExZWFiZTk3ZTRlODY2M2M4NjVkNDg4NjU3M2Q1MzM4Yzk4MzgwZjA0YmEwMGY2NDIzMzljMjdjZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HUUPnWzqI_gxrmT7_IolQ5M_VcwHQMDT6OkRL5LLkrY"> I'm not sure if I should change all the names to:
> 
> * Fields -> AdminFormFields
> * HeaderAndInstructions -> AdminFormHeaderAndInstructions (though this is quite long)
> * Logic -> AdminFormLogic
> * ThankYou -> AdminFormThankYou
> 
> I believe it's clearer, but its just quite lengthy.
> 
> ## Purpose of `admin-form/index.ts` file
> In all of the translation directories (e.g. common, login, public-form), the index.ts file is purposed as a **point of re-export** and to **define the translation interface** as such:
> 
> <img alt="image" width="465" src="https://private-user-images.githubusercontent.com/69101594/348060512-4cef129f-7b0c-4fc6-9f23-5742a8ba0a3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExMzAyOTksIm5iZiI6MTcyMTEyOTk5OSwicGF0aCI6Ii82OTEwMTU5NC8zNDgwNjA1MTItNGNlZjEyOWYtN2IwYy00ZmM2LTlmMjMtNTc0MmE4YmEwYTNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE2VDExMzk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3ZWI4YTNiODk0OThlMjk1MjA0NDA4N2Y5ZjcxYTY0NjM5NmNmYTU3NDE4OTY0OWNlMzU3ODQ1N2U5NmM2NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jqxEYTScCNz2_gLS_lcdsfjVIDcDgJ5S-O2BkRcvwQ4">
> However, for my new admin-form directory, it doesn't have any translation types. It only serves to further organize the translations into `admin-form/navbar` and `admin-form/sidebar`.
> 
> **As such, do I need an `admin-form/index.ts` file? If so, what should it contain?**

LoneRifle commented:
@jednghk actually, how about introducing `admin-form/{index,en-sg,types}.ts`, which would export an object incorporating the sub components, similar to how we already have that under the root of `locales/`? That pattern can then recurse into `navbar/` and `sidebar/` too.

jednhk committed:
refactor(i18n): recursively re-export nested resources in /admin-form

jednhk committed:
refactor(i18n): re-export from feature/{*.ts}, then specific locales

jednhk committed:
fix(i18n): lint and perform minor style fixes

jednhk commented:
> @jednghk actually, how about introducing `admin-form/{index,en-sg,types}.ts`, which would export an object incorporating the sub components, similar to how we already have that under the root of `locales/`? That pattern can then recurse into `navbar/` and `sidebar/` too.

Hi @LoneRifle, I came up with two approaches to refactoring it based on your suggestion. I've included the discussion on the tradeoffs in the PR message.

The PR to the changes can be found in my fork here:
- [Approach 1 (Flatter Hierarchy)](https://github.com/jednghk/FormSG/pull/1)
- [Approach 2 (Keep nested hierarchy)](https://github.com/jednghk/FormSG/pull/2)

LoneRifle commented:
Thanks for the effort @jednghk . To sharpen the discussion a little, I'm including the note you left in the two PRs below:
> Approach 1 is much cleaner, but unlike Approach 2, it does not reflect the hierarchy defined in types.ts

What principle or principles from Object-Oriented Programming do you think could be applicable in determining the approach you take here? Also, what other non-technical considerations do you think would be useful to think about here?

jednhk commented:
> What principle or principles from Object-Oriented Programming do you think could be applicable in determining the approach you take here?

@LoneRifle At first glance, it seems to boil down to a tradeoff between the simplicity of interface and clarity of structure.

## For Approach 1
The one that stands out to me is **Encapsulation** i.e. bundling related components under one class (potentially an AdminForm class..?) which abstracts out the complexity of the hierarchy. 

As the project scales, I can foresee abstracting the nested translation for the other features, not just for the admin-form. For instance, looking at `/public-form` has plenty more translations to add in.

## For Approach 2
I could argue that I would prefer exposing the hierarchy for the following reasons:
1. Exposing the hierarchy could be beneficial in understanding the translation components and their locations in the hierarchy, and conforms to the Translation interface defined in types.ts. 
2. It's easier to create the translation hooks, as we can follow the nested references to the translations. e.g. for `admin-form/sidebar/fields` translations, approach 2 would allow developers to follow the hierarchy easily, while approach 1 would require clicking through nested directories.

## Conclusion
If it were up to me, I'd go with approach 2 for the following reasons:
1. From a dev perspective, I like the clarity it offers in giving me an overview of where each translation file is located, which approach 1 hides. I hope that other engineers seeing the code feels the same. 
2. I feel we could be optimizing the structure too early on if we abstract it now, and the need to abstract would be much clearer as the hierarchy grows.

> Also, what other non-technical considerations do you think would be useful to think about here?

Since we're still in the process of adding translation files, exposing hierarchy is useful for future developers to see what's the progress for the translations.

I'd love to hear your thoughts on this as well!

LoneRifle commented:
@jednhk thanks for sharing! We are anticipating the scaling up of translations as we go along, in fact, this very PR is an indication of that, since it replaces a single file with a group of files covering a specific domain. This would mean that it makes sense to anticipate the abstraction of locale modules _sooner_, rather than later. 

I'd imagine that the tooling we have in place would allow the developer to figure out which translation key to use without having to expand and navigate the nested hierarchy of locales. Specifically, this tooling would be the types that would be picked up auto-complete features in VSCode and other IDEs. 

That said, I take your point that some things should be explicit. I would hence propose that while the locale files (eg, `en-sg.ts`) themselves are implicit, the types (ie, `types.ts`) should not. This should give us best of both: locale files are concise and don't have too many import statements, while the types explicitly indicate what the hierarchy of translations (and hence their components) should look like.

https://github.com/opengovsg/FormSG/blob/e23df7b461a9ced8ce071d51c8f7825d71139315/frontend/src/i18n/locales/types.ts#L6-L13

The overall goal here is to give people some sense of structure. As it is with the above block, it's not clear where each feature actually is in the interface. Header and Fields could be anywhere, there are a lot of sections in FormSG with such things.

What do you think?

jednhk commented:
> I would hence propose that while the locale files (eg, `en-sg.ts`) themselves are implicit, the types (ie, `types.ts`) should not.

@LoneRifle This addresses my main concern on exposing the hierarchy. Thanks for pointing that out.

> This should give us best of both: locale files are concise and don't have too many import statements, while the types explicitly indicate what the hierarchy of translations (and hence their components) should look like.

This makes sense. As the number of translations scale, the clutter of nesting could throw off developers more than it helps.

Keeping the `en-sg.ts` file concise with the higher-level directories would be a good compromise, while we keep a reference of the explicit hierarchy in `types.ts`.

With that said, shall I go ahead and merge [Approach 1 (flatter hierarchy)](https://github.com/jednhk/FormSG/pull/1/files)? Then we'll iterate on that. 

The key snippets can be found below:

Abstracted hierarchy in `en-sg.ts`
```ts
export const enSG: FallbackTranslation = {
  translation: {
    features: {
      adminForm,
      common,
      login,
      publicForm,
    },
  },
}
```

Exposed hierarchy in `types.ts`
```ts
interface Translation {
  translation: {
    features: {
      adminForm?: {
        sidebar?: {
          fields?: Fields
          headerAndInstructions?: HeaderAndInstructions
          logic?: Logic
          thankYou?: ThankYou
        }
        navbar?: Navbar
      }
      common?: Common
      publicForm?: PublicForm
      login?: Login
    }
  }
}
```

jednhk committed:
refactor(i18n): let FallbackTranslation use DeepRequired utility class

KenLSM commented:
Hey @jednhk there seemed to be a few frontend lint issues in https://github.com/opengovsg/FormSG/actions/runs/10155495087/job/28082557465?pr=7509. Could you take a look and address them?

jednhk commented:
> Hey @jednhk there seemed to be a few frontend lint issues in https://github.com/opengovsg/FormSG/actions/runs/10155495087/job/28082557465?pr=7509. Could you take a look and address them?

Hi Ken, thanks for taking a look.

I'm not done replacing the hardcoded English strings with translation hooks.

I'll mark the PR as a draft to prevent confusion. When it's ready for review I'll ping you.

jednhk committed:
chore(i18n): replace hardcoded text with t hooks in multiple locations

jednhk commented:
@KenLSM Just in case you missed this, it's ready for review!

LoneRifle commented:
> @KenLSM Just in case you missed this, it's ready for review!

@jednhk could you rebase or otherwise rework the branch so that you resolve the merge conflicts? Thanks!

jednhk committed:
fix(i18n): amend new translations

jednhk commented:
@LoneRifle Rebased and resolved the conflicts! 

Its my first time rebasing, do let me know if it was done correctly.

LoneRifle commented:
Yea, that doesn't look right at all. If you were to look at the files listed in your PR, they now include a bunch of files that were not part of your original contribution. Further, the git network [graph](https://github.com/opengovsg/FormSG/network) shows where the deviation is happening:
<img width="895" alt="image" src="https://github.com/user-attachments/assets/7b856b33-f92a-441d-bdec-558a12812b4b">

In future, try to avoid using the mainline branch (in this case `develop`) for development purposes unless you are highly confident that you can maintain parity with the equivalent branch of the repo you are forking from. Keep that branch solely for syncing changes with the upstream repo, then work off feature branches. This makes it much easier to rebase your changes for your PRs.

For now, you may wish to repair your branch as follows:
```
# Rebase interactively from 3f5b127, dropping all commits but 57324b2 and c2b62da
> git rebase -i 3f5b127
# Force push to the branch
> git push -f
# Perform a merge from current develop. Left as an exercise to the reader.

```

jednhk commented:
@LoneRifle As per your advice, I've rebased it to include only my own commits, and force pushed it to the repo.

I then merged the upstream repo into my own and resolved the merge conflicts. Now it only contains my own commits

It should be good now! It did seem strange that I had commits from other authors in my PR. I'll go read up more about rebasing.

Many thanks for the guidance, I'll be sure to work on a feature branch in the future when submitting a PR.

LoneRifle commented:
@KenLSM - if you could have a look?

jednhk commented:
> @KenLSM - if you could have a look?

@LoneRifle I had a look at the Chromatic review on the UI, there's some unexpected kinks in the UI I need to iron out before it's ready for review.

@KenLSM, will ping when I'm done.

I'd like to avoid this issue in the future (i.e. missing out on checking all the components thoroughly). Do you have some advice? Or perhaps an automated tool I could use? I'm not quite versed with the existing CICD technologies.

I just learnt about Storybook and looked into it a bit deeper. I know how to run it but it doesn't seem to have that same UI diff comparison that Chromatic has.

Is it possible to run a similar Chromatic comparison on my local device? Though looking into Chromatic, it seems the snapshots are paid, so it's not free at all.

Please advise. In the meantime, I'll make the necessary changes and update here.

KenLSM commented:
Sorry @jednhk our team capacity is still a little tight, the review will take a bit longer!

LoneRifle commented on frontend/src/components/Field/Attachment/Attachment.tsx:
```diff
@@ -339,12 +339,7 @@ export const Attachment = forwardRef<AttachmentProps, 'div'>(
-                  readableMaxSize,
-                },
-              )}
+              Maximum file size: {readableMaxSize}
```
Hang on, isn't this a reversion?

jednhk replied:
It's indeed a reversion. Reason being, the UI Tests on Storybook/Chromatic kept showing `{readableMaxSize}`, rather than replacing it as such:

<img width="848" alt="image" src="https://github.com/user-attachments/assets/ce185856-202c-4b00-a970-90caf6b0a4db">

Unfortunately, this value is used in many of the UI test pages, resulting in quite a number of UI diff issues raised by Chromatic.

To reduce the workload for the code review, I reverted it so we can push most of the generally non-problematic translations, as well as the new file structure for the translations. This would allow others to work on the new structure as well. After which we could make a more incremental PR for this problem.

What are your thoughts on this? Should I not have reverted it and let the team take a look?

LoneRifle replied:
I think that's fine then.

LoneRifle commented:
@jednhk - apologies, could I trouble you to keep your branch up-to-date yet again? 

@KenLSM - I would like to have this PR prioritised, in light of some recent focus on internationalisation, details of which I will discuss with you separately.

jednhk commented:
> @jednhk - apologies, could I trouble you to keep your branch up-to-date yet again?
> 
> @KenLSM - I would like to have this PR prioritised, in light of some recent focus on internationalisation, details of which I will discuss with you separately.

Sure, I'll find some time to resolve the conflicts this by the end of this week.

LoneRifle commented:
@jednhk - just some feedback - generally, try to observe the commit message conventions of the repo you're contributing to. For this particular PR, I took the trouble while rebasing to reword your commit messages to follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/), which we adhere to.

LoneRifle committed:
fix(i18n): correct typo

LoneRifle committed:
fix(i18n): add missing types

LoneRifle reviewed (approved):
Approved with changes. Note the following:

- Although some of the i18n resources have been made for the logic module, they have not actually been used. We presume that this is to minimise the number of deviations flagged by Chromatic for ease of review. These will have to be followed up at a later point.

LoneRifle committed:
fix(i18n): simplify attachment field interpolation

LoneRifle committed:
fix(i18n): avoid using plurals for verified fields

There appears to be something wrong with i18next's implementation of
plurals, which no longer seem to honour count and suffix keys

LoneRifle commented:
Duplicate of #7815
